feat(routing): reconciles routing resources

bartoszmajsak opened the pull request:
## Summary

This update introduces new logic for creating and managing routing resources based on the `routing.opendatahub.io/export-mode` annotation present on the Component's CR. This annotation can have values `public`, `external`, or a combination of `public;external`, each resulting in different resources being created.

## Internal Routing for Non-Mesh Clients (public mode)

For non-mesh clients within the cluster, the following resources are required within the `gateway namespace`:

- **Service**: Represents each internal service for non-mesh clients, named `some-service-some-namespace.gateway-ns`
- **Secret**: Contains the serving certificate and key for the Service, generated by OpenShift via an annotation on the Service.
- **Gateway**: Configures the listener for each Service and references the Secret with the serving certificate and key. The Gateway is set to only listen for the FQDN of the Service (e.g., `some-service.some-namespace.svc` and `some-service.some-namespace.svc.cluster.local`).
- **VirtualService**: Applies to both the Gateway and the mesh, routing traffic to the appropriate backend service.

## External Routing (external mode)

For services exposed externally, the following resources are created within the `gateway namespace`:

- **Route**: Uses TLS re-encrypt and targets the shared Service defined above.
- **VirtualService**: Configures the Gateway to specify the target for the Route, directing traffic to the appropriate backend service. This VirtualService does not apply to the mesh.

## Metadata for Component Services

Services that should be exposed are expected to have certain **labels** set (e.g. by their controller).

To indicate that the service should be included in platform routing `routing.opendatahub.io/exported: "true"` should be added.

 To enable garbage collection (by this controller, not k8s itself) for platform routing resources, the following labels should be added:

- `platform.opendatahub.io/owner-name`: contains the owning Component's CR name.
- `platform.opendatahub.io/owner-kind`: contains the owning Component's CR kind.

These labels are needed because we cannot use  `OwnerReference` in all cases. Most components have their Custom Resources (CRs) namespace-scoped, as cross-namespace owner references are disallowed by design, we cannot use them as `OwnerReference`  for routing resources that we need to create in the `GATEWAY_NAMESPACE`. 

## Updating host information

When new routing resources are successfully created and new hosts defined, the routing controller places the following annotations directly on the Component CR instead of individual Services:

* `routing.opendatahub.io/public-addresses`
* `routing.opendatahub.io/external-addresses`

Both have values in a format of `;` separated list, e.g. 

`routing.opendatahub.io/addresses-public: some-service-some-namespace.gateway-ns;some-service-some-namespace.gateway-ns.svc;some-service-some-namespace.gateway-ns.svc.cluster.local`

This enables Platform Auth to read these addresses from the Component CR and update `AuthConfig`s host definitions accordingly.

> [!IMPORTANT]
> Implementation of `AuthConfig`s update pending.

## Current limitations

- Each component service to be exposed must have only one port defined.
- There is no support for VirtualService delegation yet. Meaning `VirtualService` with `routing.opendatahub.io/exported: "true"` label is not handled by the controller.
- Created resources are not automatically removed upon deletion of a watched CR (WIP follow-up issue: [RHOAIENG-10727](https://issues.redhat.com/browse/RHOAIENG-10727)).
- Resources required for a given export mode are not removed if the watched CR changes to a different export mode (follow-up issue: [RHOAIENG-11030](https://issues.redhat.com/browse/RHOAIENG-11030)).

## Additional Changes

### Reading Default Ingress Domain

This update ensures that externally exposed services have routes with hosts based on the public service name and domain configured for the OpenShift Ingress controller.

The external CRD for `ingresses.config.openshift.io` is now fetched as part of `make generate` using `$(call fetch-external-crds,github.com/openshift/api,config/v1)`. Notably, there is an inconsistency with the added CRD files, as they differ from those defined in the repository. Specifically, the Ingress CRD is mistakenly set to namespace-scoped instead of the expected cluster-scoped.


## Fixes

* https://issues.redhat.com/browse/RHOAIENG-10702
* https://issues.redhat.com/browse/RHOAIENG-10732

bartoszmajsak committed:
feat(routing): reconciles routing resources

This update introduces new logic for creating and managing routing resources based on the `routing.opendatahub.io/export-mode` annotation. This annotation can have values `public`, `external`, or a combination of `public;external`, each resulting in different resources being created.

Internal Routing for Non-Mesh Clients (public mode)

For non-mesh clients within the cluster, the following resources are required within the `gateway namespace`:

- **Service**: Represents each internal service for non-mesh clients.
- **Secret**: Contains the serving certificate and key for the Service, generated by OpenShift via an annotation on the Service.
- **Gateway**: Configures the listener for each Service and references the Secret with the serving certificate and key. The Gateway is set to only listen for the FQDN of the Service (e.g., `some-service.some-namespace.svc` and `some-service.some-namespace.svc.cluster.local`).
- **VirtualService**: Applies to both the Gateway and the mesh, routing traffic to the appropriate backend service.

External Routing (external mode)

For services exposed externally, the following resources are created within the `gateway namespace`:

- **Route**: Uses TLS re-encrypt and targets the shared Service defined above.
- **VirtualService**: Configures the Gateway to specify the target for the Route, directing traffic to the appropriate backend service. This VirtualService does not apply to the mesh.

Annotations for Component Services

Services that should be exposed are expected to have the following metadata set by their controller:

- `routing.opendatahub.io/exported: "true"`: Indicates the service should be included in platform routing.
- `platform.opendatahub.io/owner-name`: Contains the owning Component's CR name.
- `platform.opendatahub.io/owner-kind`: Contains the owning Component's CR kind.

The routing controller places annotations directly on the Component CR instead of individual Services. This ensures the Platform Router can write back exposed addresses and that Platform Auth can read these addresses from the Component CR.

Limitations and Assumptions

- Each component service to be exposed must have only one port defined.
- There is no support for VirtualService delegation.
- Created resources are not automatically removed upon deletion of a watched CR ([RHOAIENG-10727](https://issues.redhat.com/browse/RHOAIENG-10727)).
- Resources required for a given export mode are not removed if the watched CR changes to a different export mode ([RHOAIENG-11030](https://issues.redhat.com/browse/RHOAIENG-11030)).

Additional Changes

**Reading Default Ingress Domain**

This update ensures that externally exposed services have routes with hosts based on the public service name and domain configured for the OpenShift Ingress controller.

The external CRD for `ingresses.config.openshift.io` is now fetched as part of `make generate` using `$(call fetch-external-crds,github.com/openshift/api,config/v1)`. Notably, there is an inconsistency with the added CRD files, as they differ from those defined in the repository. Specifically, the Ingress CRD is mistakenly set to namespace-scoped instead of the expected cluster-scoped.

cam-garrison commented on controllers/routing/exported_svc_locator.go:
```diff
@@ -0,0 +1,38 @@
+		return nil, &NoExportedServicesError{Target: target}
+	}
+
+	return &exportedSvcList.Items[0], nil
```
So we are (for now) limiting to one exported service per namespace?

bartoszmajsak replied:
Good catch. We can easily handle more.

bartoszmajsak replied:
Done in 30c0a49

cam-garrison reviewed (commented):
added question

cam-garrison commented on controllers/routing/routing_controller_test.go:
```diff
@@ -0,0 +1,578 @@
+				WithTimeout(test.DefaultTimeout).
+				WithPolling(test.DefaultPolling).
+				Should(Succeed())
+
```
worth adding one for destination rule too? and above in the public only test?

bartoszmajsak replied:
Added in d5da2e6

bartoszmajsak committed:
feat: handles routing for multiple svc per target

Additionally wraps the call in retry in case SVCs are not yet created in the cluster when watched CR triggers reconcile on our end.

bartoszmajsak committed:
chore: simplifies routing template data struct

by composing it with PlatformRoutingConfiguration struct

bartoszmajsak committed:
feat(test): adds verification for destination policy

bartoszmajsak committed:
chore: renames custom error

cam-garrison commented:
in `pkg/metadata/label.go` 

comment:
	// OwnerName is the name of the owner of the resource combined with its namespace separated by a dash if applicable.
	
exists but this piece re: including namespace is not yet covered

bartoszmajsak committed:
chore(meta): renames routing annotations

bartoszmajsak committed:
chore(test): reworks gomega matchers

Removes unneeded custom matchers where asserting on actual field, such as .Annotations map, can be used in conjuction with built-in matchers. Previous approach did not improve readability of the test much but more importantly made error reporting cumbersome by hiding actual mismatch details which built-in matchers have out of the box

bartoszmajsak committed:
chore(test): moves resources test fixture to the bottom of the file

bartoszmajsak committed:
chore(test): improves error msgs

By relying on actual built-in matcher used

bartoszmajsak committed:
chore(test): simplifies test structure

By extracting common part describing requirements for a component
to become part of platform routing it becomes easier to understand
what are the requirements for a component to be part of platform
routing. These pieces were redundant and not easy to immediately graps
as there was quite some boilerplate code overshadowing the actual
narrative.

bartoszmajsak committed:
feat(routing): propagates owner UID through the label

Having this metadata will allow us to identify resources required deletion when
owning component is no longer existing. Labels are to address following limitations:

- owning resource is namespace-scoped and exists in different namespace
  (owner-name and owner-kind are sufficient)
- there is more than one resource of the same kind and name, but in
  different namespaces (owner-uid allows to identify exact ownership)

bartoszmajsak commented:
> in `pkg/metadata/label.go`
> 
> comment: // OwnerName is the name of the owner of the resource combined with its namespace separated by a dash if applicable.
> 
> exists but this piece re: including namespace is not yet covered

@cam-garrison I added UID propagation 70daed6, you can base your work on it.

cam-garrison commented on controllers/routing/exported_svc_locator.go:
```diff
@@ -0,0 +1,62 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func getExportedServices(ctx context.Context, cli client.Client, target *unstructured.Unstructured) ([]corev1.Service, error) {
+	ownerName := target.GetName()
+	ownerKind := target.GetObjectKind().GroupVersionKind().Kind
+
+	listOpts := []client.ListOption{
+		client.InNamespace(target.GetNamespace()),
+		// TODO(mvp): centralize label creation
+		client.MatchingLabels{
+			metadata.Labels.RoutingExported: "true",
+			metadata.Labels.OwnerName:       ownerName,
+			metadata.Labels.OwnerKind:       ownerKind,
+		},
+	}
```
worth adding UID label matching here? 

follow up question (maybe sth for future iteration past MVP) - not sure if `client.InNamespace(target.GetNamespace()),` will work for cluster scoped resources? worth checking at some pt.

bartoszmajsak replied:
> worth adding UID label matching here?

We don't need to, UID is our internal metadata and SVC is namespace-scoped there is no risk in duplication here.

> follow up question (maybe sth for future iteration past MVP) - not sure if client.InNamespace(target.GetNamespace()), will work for cluster scoped resources? worth checking at some pt.

In general, it won't work, but `Service` (and `VirtualService` later too) which we are exporting are namespaced-scope resources, so all good for this case.

cam-garrison reviewed (approved):
left one comment, but LGTM :tada:
